pacai/agents/search/base.py

```diff
@@ -16,18 +19,29 @@ class SearchAgent(BaseAgent):
     """
 
     def __init__(self, index,
-            fn = 'pacai.student.search.depthFirstSearch',
-            prob = 'pacai.core.search.position.PositionSearchProblem',
-            heuristic = 'pacai.core.search.heuristic.null',
+            fn: Union[str, Callable[[SearchProblem], any]]
+                = 'pacai.student.search.depthFirstSearch',
+            prob: Union[str, Callable[[AbstractGameState], SearchProblem]]
+                  = 'pacai.core.search.position.PositionSearchProblem',
+            heuristic: str = 'pacai.core.search.heuristic.null',
             **kwargs):
         super().__init__(index, **kwargs)
 
-        # Get the search problem type from the name.
-        self.searchType = reflection.qualifiedImport(prob)
+        if isinstance(prob, str):
+            # Get the search problem type from the name.
+            self.searchType = reflection.qualifiedImport(prob)
+        else:
+            # Use provided problem or problem factory
+            self.searchType = prob
         logging.info('[SearchAgent] using problem type %s.' % (prob))
```

Comment:
Log `self.searchType` instead of `prob`.
(If you want you can log both, either in the same log or log `prob` in a debug log after.)